src/conn/conn_compact.c

```diff
@@ -21,21 +22,256 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
+    TAILQ_INSERT_HEAD(&(conn)->background_compact.compacthash[bucket], compact_stat, hashq);
+    ++(conn)->background_compact.file_count;
+}
+
+/*
+ * __background_compact_list_remove --
+ *     Remove compaction statistics for a file from the background compact list.
+ */
+static void
+__background_compact_list_remove(WT_SESSION_IMPL *session, WT_BACKGROUND_COMPACT_STAT *compact_stat)
+{
+    WT_CONNECTION_IMPL *conn;
+    uint64_t bucket, hash;
+
+    conn = S2C(session);
+
+    hash = __wt_hash_city64(compact_stat->uri, strlen(compact_stat->uri));
+    bucket = hash & (conn->hash_size - 1);
+
+    TAILQ_REMOVE(&(conn)->background_compact.compactqh, compact_stat, q);
+    TAILQ_REMOVE(&(conn)->background_compact.compacthash[bucket], compact_stat, hashq);
+    WT_ASSERT(session, (conn)->background_compact.file_count > 0);
+    --(conn)->background_compact.file_count;
+
+    __wt_free(session, compact_stat->uri);
+    __wt_free(session, compact_stat);
+}
+
+/*
+ * __background_compact_get_stat --
+ *     Get the statistics for the given uri.
+ */
+static WT_BACKGROUND_COMPACT_STAT *
+__background_compact_get_stat(WT_SESSION_IMPL *session, const char *uri, int64_t id)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat, *temp_compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t bucket, hash;
+
+    conn = S2C(session);
+
+    hash = __wt_hash_city64(uri, strlen(uri));
+    bucket = hash & (conn->hash_size - 1);
+
+    /* Find the uri in the files compacted list. */
+    TAILQ_FOREACH_SAFE(
+      compact_stat, &conn->background_compact.compacthash[bucket], hashq, temp_compact_stat)
+    if (strcmp(uri, compact_stat->uri) == 0) {
+        /*
+         * If we've found an entry in the list with the same URI but different ID's we must've
+         * dropped and recreated this table. Reset the entry in this case.
+         */
+        if (id != compact_stat->id) {
+            __background_compact_list_remove(session, compact_stat);
+            return (NULL);
+        }
+
+        return (compact_stat);
+    }
+
+    return (NULL);
+}
+
+/*
+ * __background_compact_should_run --
+ *     Check whether we should proceed with calling compaction on the given file.
+ */
+static bool
+__background_compact_should_run(WT_SESSION_IMPL *session, const char *uri, int64_t id)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t cur_time;
+
+    conn = S2C(session);
+
+    /* The history store file should not be compacted. */
+    if (WT_STREQ(uri, WT_HS_URI))
+        return (false);
+
+    /* If we haven't seen this file before we should try and compact it. */
+    compact_stat = __background_compact_get_stat(session, uri, id);
+    if (compact_stat == NULL)
+        return (true);
+
+    /* If we have been unsuccessful recently skip this file for some time. */
+    cur_time = __wt_clock(session);
+    if (!compact_stat->prev_compact_success &&
+      WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) <
```

Comment:
I'd find this easier to read if the time check was pulled out into a first condition:

```
if (WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) >    WT_BACKGROUND_COMPACT_SLEEP_TIME)
    return (true);
```

Then the two conditions could be very simple. If you take the naming suggestion for the constant above, this would change as well, but you get the idea.

Reply:
Nice, yeah that's more readable.

Reply:
Fixed https://github.com/wiredtiger/wiredtiger/pull/9541/commits/112ccd9c97af7114df00e6b89c772d59b5d29444.